File: src/classes/BitBuffer.js
class BitBuffer {
    constructor() {
        this.bits = [];
    }

    _appendBits(value, bitCount) {
        for (let i = bitCount - 1; i >= 0; i--) {
            const bit = (value >> i) & 1;
            this.bits.push(bit);
        }
    }

    writeUtfString(text) {
        if (text === null || text === undefined) {
            text = "";
        }
        const length = Buffer.byteLength(text, "utf8");
        this._appendBits((length >> 8) & 0xff, 8);
        this._appendBits(length & 0xff, 8);
        for (let i = 0; i < text.length; i++) {
            this._appendBits(text.charCodeAt(i) & 0xff, 8);
        }
    }

    writeMethod4(val) {
        if (val === 0) {
            const loc1 = 0;
            const loc2 = 2;
            this._appendBits(loc1, 4);
            this._appendBits(val, loc2);
            return;
        }
        let n = Math.floor(Math.log2(val)) + 1;
        if (n % 2 !== 0) {
            n += 1;
        }
        const loc1 = (n >> 1) - 1;
        this._appendBits(loc1, 4);
        this._appendBits(val, n);
    }

    writeMethod393(val) {
        this._appendBits(val & 0xff, 8);
    }

    writeMethod6(val, bitCount) {
        this._appendBits(val, bitCount);
    }

    writeMethod13String(text) {
        const length = text.length;
        this.writeMethod4(length);
        for (let i = 0; i < text.length; i++) {
            this._appendBits(text.charCodeAt(i) & 0xff, 8);
        }
    }

    toBytes() {
        while (this.bits.length % 8 !== 0) {
            this.bits.push(0);
        }
        const out = Buffer.alloc(Math.ceil(this.bits.length / 8));
        for (let i = 0; i < this.bits.length; i += 8) {
            let b = 0;
            for (let j = 0; j < 8 && i + j < this.bits.length; j++) {
                b = (b << 1) | this.bits[i + j];
            }
            out[i / 8] = b;
        }
        return out;
    }
}

export { BitBuffer };

File: src/classes/bit_buffer.py
class BitBuffer:
    def __init__(self):
        self.bits = []

    def _append_bits(self, value, bit_count):
        for i in reversed(range(bit_count)):
            bit = (value >> i) & 1
            self.bits.append(bit)

    def write_utf_string(self, text):
        if text is None:
            text = ""
        length = len(text)
        self._append_bits((length >> 8) & 0xFF, 8)
        self._append_bits(length & 0xFF, 8)
        for ch in text:
            self._append_bits(ord(ch) & 0xFF, 8)

    def write_method_4(self, val):
        if val == 0:
            _loc1_ = 0
            _loc2_ = 2
            self._append_bits(_loc1_, 4)
            self._append_bits(val, _loc2_)
            return
        n = val.bit_length()
        if n % 2 != 0:
            n += 1
        _loc1_ = (n >> 1) - 1
        self._append_bits(_loc1_, 4)
        self._append_bits(val, n)

    def write_method_393(self, val):
        self._append_bits(val & 0xFF, 8)

    def write_method_6(self, val, bit_count):
        self._append_bits(val, bit_count)

    def write_method_13_string(self, text):
        length = len(text)
        self.write_method_4(length)
        for ch in text:
            self._append_bits(ord(ch) & 0xFF, 8)

    def to_bytes(self):
        while len(self.bits) % 8 != 0:
            self.bits.append(0)
        out = bytearray()
        for i in range(0, len(self.bits), 8):
            b = 0
            for bit in self.bits[i:i+8]:
                b = (b << 1) | bit
            out.append(b)
        return bytes(out)